File: app/src/main/java/ua/com/foxminded/locationtrackera/di/AppModule.java
package ua.com.foxminded.locationtrackera.di;

import com.google.firebase.auth.FirebaseAuth;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

import ua.com.foxminded.locationtrackera.App;
import ua.com.foxminded.locationtrackera.model.auth.AuthNetwork;
import ua.com.foxminded.locationtrackera.model.auth.FirebaseAuthNetwork;
import ua.com.foxminded.locationtrackera.model.locations.LocationRepository;
import ua.com.foxminded.locationtrackera.model.shared_preferences.DefaultSharedPreferencesModel;
import ua.com.foxminded.locationtrackera.model.shared_preferences.SharedPreferencesModel;
import ua.com.foxminded.locationtrackera.model.usecase.SendLocationsUseCase;
import ua.com.foxminded.locationtrackera.model.usecase.SendLocationsUseCaseImpl;

@Module
public class AppModule {

    @Provides
    @Singleton
    public AuthNetwork provideFirebaseAuthNetwork() {
        return new FirebaseAuthNetwork(FirebaseAuth.getInstance());
    }

    @Provides
    public SendLocationsUseCase provideSendLocationsUseCase(LocationRepository repository) {
        return new SendLocationsUseCaseImpl(repository);
    }

    @Provides
    public SharedPreferencesModel provideSharedPreferenceModel() {
        return new DefaultSharedPreferencesModel(App.getInstance());
    }
}

File: app/src/main/java/ua/com/foxminded/locationtrackera/di/AppModule.kt
package ua.com.foxminded.locationtrackera.di

import com.google.firebase.auth.FirebaseAuth
import dagger.Module
import dagger.Provides
import ua.com.foxminded.locationtrackera.models.auth.AuthNetwork
import ua.com.foxminded.locationtrackera.models.locations.LocationRepository
import ua.com.foxminded.locationtrackera.models.usecase.SendLocationsUseCase
import ua.com.foxminded.locationtrackera.models_impl.auth.FirebaseAuthNetwork
import ua.com.foxminded.locationtrackera.models_impl.usecase.SendLocationsUseCaseImpl
import javax.inject.Singleton

@Module
class AppModule {

    @Provides
    @Singleton
    fun provideFirebaseAuthNetwork(): AuthNetwork = FirebaseAuthNetwork(FirebaseAuth.getInstance())

    @Provides
    fun provideSendLocationsUseCase(repository: LocationRepository): SendLocationsUseCase =
        SendLocationsUseCaseImpl(repository)
}